File: clients/rust/turk-router/tests/common/harness/meteora_damm_v2.rs
use solana_pubkey::Pubkey;
use turk_router::venues::meteora_damm_v2::{
    resolve as build_window, DammV2Form, MeteoraDammV2Accounts, PROGRAM_ID,
};
use turk_router::VenueWindow;

use crate::common::fixture::Fixture;

pub fn resolve(fixture: &Fixture) -> VenueWindow {
    assert_eq!(fixture.program_id, PROGRAM_ID);

    let referral_slot = fixture.slot(11).pubkey;
    let accounts = MeteoraDammV2Accounts {
        pool: fixture.slot(1).pubkey,
        input_token_account: fixture.slot(2).pubkey,
        output_token_account: fixture.slot(3).pubkey,
        token_a_vault: fixture.slot(4).pubkey,
        token_b_vault: fixture.slot(5).pubkey,
        token_a_mint: fixture.slot(6).pubkey,
        token_b_mint: fixture.slot(7).pubkey,
        payer: fixture.slot(8).pubkey,
        token_a_program: fixture.slot(9).pubkey,
        token_b_program: fixture.slot(10).pubkey,
        referral_token_account: (referral_slot != PROGRAM_ID).then_some(referral_slot),
    };

    let form = match fixture.slots.len() {
        14 => DammV2Form::Base,
        15 => DammV2Form::RateLimited,
        other => panic!(
            "{}: unexpected meteora damm v2 slot count {other}",
            fixture.pool_b58
        ),
    };

    build_window(accounts, form)
}

pub fn reachable_account_counts() -> Vec<u8> {
    fn key(byte: u8) -> Pubkey {
        Pubkey::new_from_array([byte; 32])
    }

    fn accounts(referral_token_account: Option<Pubkey>) -> MeteoraDammV2Accounts {
        MeteoraDammV2Accounts {
            pool: key(1),
            input_token_account: key(2),
            output_token_account: key(3),
            token_a_vault: key(4),
            token_b_vault: key(5),
            token_a_mint: key(6),
            token_b_mint: key(7),
            payer: key(8),
            token_a_program: key(9),
            token_b_program: key(10),
            referral_token_account,
        }
    }

    let mut counts = Vec::new();
    for form in [DammV2Form::Base, DammV2Form::RateLimited] {
        for referral_token_account in [None, Some(key(11))] {
            counts.push(build_window(accounts(referral_token_account), form).account_count());
        }
    }
    counts
}

File: clients/ts/tests/common/harness/meteoraDammV2.ts
import type { Address } from "@solana/addresses";

import { venues, type VenueWindow } from "../../../src/index.js";
import { slotAt, type Fixture } from "../fixture.js";
import { key } from "../keys.js";
import { assertProgram } from "./index.js";

const { DammV2Form, PROGRAM_ADDRESS, resolve: build } = venues.meteoraDammV2;

export function resolve(fixture: Fixture): VenueWindow {
  assertProgram(fixture, PROGRAM_ADDRESS);
  const referralSlot = slotAt(fixture, 11).pubkey;
  const accounts: venues.meteoraDammV2.MeteoraDammV2Accounts = {
    pool: slotAt(fixture, 1).pubkey,
    inputTokenAccount: slotAt(fixture, 2).pubkey,
    outputTokenAccount: slotAt(fixture, 3).pubkey,
    tokenAVault: slotAt(fixture, 4).pubkey,
    tokenBVault: slotAt(fixture, 5).pubkey,
    tokenAMint: slotAt(fixture, 6).pubkey,
    tokenBMint: slotAt(fixture, 7).pubkey,
    payer: slotAt(fixture, 8).pubkey,
    tokenAProgram: slotAt(fixture, 9).pubkey,
    tokenBProgram: slotAt(fixture, 10).pubkey,
    referralTokenAccount: referralSlot === PROGRAM_ADDRESS ? undefined : referralSlot,
  };
  switch (fixture.slots.length) {
    case 14:
      return build(accounts, DammV2Form.Base);
    case 15:
      return build(accounts, DammV2Form.RateLimited);
    default:
      throw new Error(
        `${fixture.poolB58}: unexpected meteora damm v2 slot count ${String(fixture.slots.length)}`,
      );
  }
}

function accountsFor(
  referralTokenAccount: Address | undefined,
): venues.meteoraDammV2.MeteoraDammV2Accounts {
  return {
    pool: key(1),
    inputTokenAccount: key(2),
    outputTokenAccount: key(3),
    tokenAVault: key(4),
    tokenBVault: key(5),
    tokenAMint: key(6),
    tokenBMint: key(7),
    payer: key(8),
    tokenAProgram: key(9),
    tokenBProgram: key(10),
    referralTokenAccount,
  };
}

export function reachableAccountCounts(): Array<number> {
  const counts: Array<number> = [];
  for (const form of [DammV2Form.Base, DammV2Form.RateLimited]) {
    for (const referralTokenAccount of [undefined, key(11)]) {
      counts.push(build(accountsFor(referralTokenAccount), form).accountCount);
    }
  }
  return counts;
}
